### DISCUSSION TITLE
Mx, My and Mz  in Mellinger Controller

### DISCUSSION THREAD

**Discussion by magalhaesw** [bitcraze/discussions#645] [Q&A]
Hi All !
I have some questions about the implementation of Mellinger controller.
Please,
Could you help me?

The Mellinger paper explains that we have Thrust and Moments (Mx, My, and Mz) to control CF.
At  https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller/controller_mellinger.c#L295
We have :
```
  if (control->thrust > 0) {
    control->roll = clamp(M.x, -32000, 32000);
    control->pitch = clamp(M.y, -32000, 32000);
    control->yaw = clamp(-M.z, -32000, 32000);
```
So, 
I understand that **"control->roll"**  is the Moment in *x* (Mx) .

The term **"control->roll"** also appears in INDI controller at:
https://github.com/bitcraze/crazyflie-firmware/blob/9371e5f5036e8619374db403f9e3141a2e10aa7c/src/modules/src/controller/controller_indi.c#L330
```
/*  INDI feedback */
	control->thrust = indi.thrust;
	control->roll = indi.u_in.p;
	control->pitch = indi.u_in.q;
	control->yaw  = indi.u_in.r;
```
In this part of the code we have  **"control->roll = p".** 
Does it mean **"control->roll"** is angular velocity (in rad/s) ?

My question is: 
1) What is the meaning of  **"control->roll"?**
Is it angular velocity or Moment in x ? What is the unit for "control->roll"? 

Thank you so much for clarification.

Best,
   JM

**Comment by knmcguire** [comment]
Have you found this file yet? https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/power_distribution_quadrotor.c

It should clarify how these values are used for motor mixing
